### PR TITLE
Grades Import Pipeline

### PR DESCRIPTION
**Pull request by bradenmacdonald**
**Description**: This PR introduces a series of pipeline tasks that import grades from the LMS for display in Insights.

Specifically, the end result is that two new MySQL tables are created in the analytics DB:
- `grades`: For each course with grades available, this table gives the letter grade, percentage grade, and whether or not the student is passing for every student. This table is intended to be displayed on [the new student profile pages](https://github.com/edx/edx-analytics-dashboard/pull/338) in Insights:
  user profile mockup
  _(don't pay too much attention to the UX above - this is a pipeline PR, not a dashboard/UX PR)_
- `grade_breakdown`: For each course with grades available, this table contains the distribution of letter grades as a function of time. This table is intended to support a new graph in the Insights Dashboard that shows the % of students with a passing grade over time, as a percentage or an absolute number.

Because [grading is slow](https://openedx.atlassian.net/wiki/display/TNL/edX+Grading+Guide+for+Developers), this new pipeline only imports **offline computed grades**. Thus, an instance wishing to use this feature needs to schedule the offline computation of grades to happen [daily] before the analytics pipeline runs. The offline grade computation should be able to cover all courses on small and medium Open edX instances, and could probably be enabled for selected smaller courses on a large instance like edx.org.

**Dependencies**:
- https://github.com/edx/edx-platform/pull/9330 (Fix Offline Grade Calculation)

**Partner information**: not an edX partner - 3rd party-hosted open edX instance

**Test Instructions**:
- In Studio, create a course with a few graded problems. Edit the course grading settings appropriately.
- In the LMS login as various users and complete the problems. Check the grades that each user earns and try to have some users with high grades and others with low grades.
- As the `edxapp` user, compute the offline grades for your course:
  
  ```
  ./manage.py lms compute_grades course-v1:Course+ID+here --settings=devstack
  ```
- Run the pipeline:
  
  ```
  export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
  remote-task --vagrant-path ../ --remote-name devstack --override-config ${PWD}/config/devstack.cfg --repo https://github.com/open-craft/edx-analytics-pipeline.git --branch grades --wheel-url $WHEEL_URL --wait GradesPipelineTask --local-scheduler --n-reduce-tasks 1
  ```
- Check the results using `mysql -u root reports` and running the queries `SELECT * FROM grades;` and `SELECT * FROM grade_breakdown;`.

To reset the pipeline:

```
mysql -u root reports -e 'drop table grade_breakdown; drop table grades; delete from table_updates where target_table IN ("grades", "grade_breakdown");'
hive --service cli -e "drop table grade_breakdown; drop table grades;"
hdfs dfs -rm -r /edx-analytics-pipeline/warehouse/{grade*,offline_grade*}
```

**Reviewers**: @itsjeyd and TBD

### CODE REVIEW

File: edx/analytics/tasks/grades.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        yield ((course_id, ), (user_id, updated, gradeset))
+
+    def reducer(self, key, entries):
+        """
+        Parse the JSON data and separate out the fields we care about, then format them to be
+        loaded as a Hive table.
+        """
+        course_id = key[0]
+
+        for user_id, _updated, gradeset_str in entries:
```
Comment (itsjeyd) [inline edx/analytics/tasks/grades.py:113]: @bradenmacdonald Why the leading underscore in `_updated`?

File: edx/analytics/tasks/grades.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        """
+        course_id = key[0]
+        grade_breakdown = {}  # key is the letter grade, value is the # of students with that grade
+        is_passing_map = {}   # key is the letter grade, value is boolean indicating if it's a passing grade
+        total_enrolled = 0    # The offline grades table has an entry for each enrolled
+                              # student, whether or not the student is active. This would only
+                              # be unreliable if the offline grade computation did not begin and
+                              # end after our 'updated' date cutoff time.
+
+        for _user_id, letter_grade, is_passing in entries:
```
Comment (itsjeyd) [inline edx/analytics/tasks/grades.py:188]: @bradenmacdonald Why the leading underscore in `_user_id`?

File: edx/analytics/tasks/grades.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+            total_enrolled += 1
+            grade_breakdown[letter_grade] = grade_breakdown.get(letter_grade, 0) + 1
```
Comment (itsjeyd) [inline edx/analytics/tasks/grades.py]: @bradenmacdonald Maybe you could shorten this a bit by using a `defaultdict`?

``` python
grade_breakdown = defaultdict(int)
# ...
    grade_breakdown[letter_grade] += 1
```

File: edx/analytics/tasks/grades.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+            ('percent_students', 'DOUBLE'),
+            ('is_passing', 'BOOLEAN'),
+        ]
+
+    @property
+    def default_columns(self):
+        """ Columns and constraints that are managed by MySQL """
+        return [
+            ('created', 'TIMESTAMP DEFAULT NOW()'),
+            ('CONSTRAINT cdg', 'UNIQUE (course_id, date, letter_grade)'),
```
Comment (itsjeyd) [inline edx/analytics/tasks/grades.py]: @bradenmacdonald Maybe you could replace `cdg` with `course_date_grade` here? That would be consistent with the per-student `grades` table (you did not abbreviate the `CONSTRAINT` for that one).

File: edx/analytics/tasks/tests/acceptance/test_grades.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+"""
+End to end test of grades import.
+"""
+
+import datetime
+
+from edx.analytics.tasks.tests.acceptance import AcceptanceTestCase
+
+
+class EnrollmentAcceptanceTest(AcceptanceTestCase):
```
Comment (itsjeyd) [inline edx/analytics/tasks/tests/acceptance/test_grades.py]: @bradenmacdonald Probably a copy-paste error :) This shouldn't be called `EnrollmentAcceptanceTest` -- how about `GradesImportAcceptanceTest`?

**Comment by itsjeyd** [comment]
@bradenmacdonald The new tasks work well (I tested with the OpenCraft devstack). Found a few small points to address, otherwise: :+1:

File: edx/analytics/tasks/grades.py (hunk 2 above)
Comment (bradenmacdonald) [inline edx/analytics/tasks/grades.py:188]: @itsjeyd The leading underscore indicates the value is not used (and prevents pylint from complaining about it).

File: edx/analytics/tasks/grades.py (hunk 3 above)
Comment (bradenmacdonald) [inline edx/analytics/tasks/grades.py]: Good idea, thanks.

**Comment by bradenmacdonald** [comment]
Thanks @itsjeyd! I pushed a commit to address your feedback. I also pushed a minor fix to the acceptance test, and I have now confirmed that the acceptance test is passing (on my devstack at least).

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-774](https://openedx.atlassian.net/browse/OSPR-774) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've added yourself to the [AUTHORS](https://github.com/edx/edx-analytics-pipeline/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

**Comment by openedx-webhooks** [comment]
Hello @bradenmacdonald: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

**Pull request closed without merging**